hikari/internal/enums.py

```diff
@@ -219,18 +210,568 @@ def __repr__(cls) -> str:
+        `builtins.ValueError` to be raised.
+
+    Operators on each enum member
+    -----------------------------
+    * `e1 == e2` : `builtins.bool`
+        Compare equality.
+    * `e1 != e2` : `builtins.bool`
+        Compare inequality.
+    * `builtins.repr(e)` : `builtins.str`
+        Get the machine readable representation of the enum member `e`.
+    * `builtins.str(e)` : `builtins.str`
+        Get the `builtins.str` name of the enum member `e`.
+
+    Special properties on each enum member
+    --------------------------------------
+    * `name` : `builtins.str`
+        The name of the member.
+    * `value` :
+        The value of the member. The type depends on the implementation type
+        of the enum you are using.
+
+    All other methods and operators on enum members are inherited from the
+    member's __value__. For example, an enum extending `builtins.int` would
+    be able to be used as an `int` type outside these overridden definitions.
+    """
+
+    _name_to_member_map_: typing.Final[typing.ClassVar[typing.Mapping[str, Enum]]]
+    _value_to_member_map_: typing.Final[typing.ClassVar[typing.Mapping[int, Enum]]]
+    _member_names_: typing.Final[typing.ClassVar[typing.Sequence[str]]]
+    __members__: typing.Final[typing.ClassVar[typing.Mapping[str, Enum]]]
+    __objtype__: typing.Final[typing.ClassVar[typing.Type[typing.Any]]]
+    __enumtype__: typing.Final[typing.ClassVar[typing.Type[Enum]]]
+    _name_: typing.Final[str]
+    _value_: typing.Final[typing.Any]
+
+    @property
+    def name(self) -> str:
+        """Return the name of the enum member as a `builtins.str`."""
+        return self._name_
+
+    @property
+    @typing.no_type_check
+    def value(self):
+        """Return the value of the enum member."""
+        return self._value_
 
     def __repr__(self) -> str:
-        return f"<{type(self).__name__}.{self.name}: {self.value!r}>"
+        return f"<{type(self).__name__}.{self._name_}: {self._value_!r}>"
 
     def __str__(self) -> str:
-        return f"{type(self).__name__}.{self.name}"
+        return self._name_ or "NO_NAME"
+
+
+_Flag = NotImplemented
+
+
+def _name_resolver(members: typing.Dict[int, _Flag], value: int) -> typing.Generator[str, typing.Any, None]:
+    bit = 1
+    has_yielded = False
+    remaining = value
+    while bit <= value:
+        if member := members.get(bit):
+            # Use ._value_ to prevent overhead of making new members each time.
+            # Also lets my testing logic for the cache size be more accurate.
+            if member._value_ & remaining == member._value_:
+                remaining ^= member._value_
+                yield member.name
+                has_yielded = True
+        bit <<= 1
+
+    if not has_yielded:
+        yield f"UNKNOWN 0x{value:x}"
+    elif remaining:
+        yield hex(remaining)
+
+
+class _FlagMeta(type):
+    def __call__(cls, value: typing.Any) -> typing.Any:
```

Comment:
Might be worth it defaulting `value` to a value that's considered empty for `_FlagMeta.__call__` (most likely just `0`) as to allow an empty call to default to no flags which would be similar to how `str()` and `int()`  both behave.

Reply:
done.